BUG: signal.resample: Fix bug for parameter num=2 (including refactor and improved docstr)

DietBru opened the pull request:
In function `resample`:
* Refactor to make code a little bit more accessible.
* Rework of docstr (also Closes #13788).
* Fix indexing bug for parameter `num=2` (Closes #14569 )
* Fix a corner case when parameter `t` is of integer type (Closes #15153)

In `signal.test.tests_signaltools.py`:
* Add unit test for verifying fix of #14569
* Increase coverage of `resample` and `resample_polyphase` to 100%.

Additionally:
* Handle unpaired bins correctly in `signal.envelope()`.
* Improve docstr of `signal.decimate` clarifying what to do with non-integer down-sampling factors (Closes #13789).

DietBru committed:
BUG: signal.resample: Fix complex-valued signal problems.

In function `resample`:
* Refactor to make code a little bit more accessible.
* Rework of docstr (also Closes #13788).
* Fix indexing bug for parameter `num=2` (Closes #14569)
* Fix a corner case when parameter `t` is of integer type (Closes #15153)

In `signal.test.tests_signaltools.py`:
* Add unit test for verifying fix of #14569
* Increase coverage of `resample` and `resample_polyphase` to 100%.

Additionally:
* Handle unpaired bins correctly in `signal.envelope()`.
* Improve docstr of `signal.decimate` clarifying what to do with non-integer down-sampling factors (Closes #13789).

DietBru commented:
Rendered docs: [resample], [decimate]

[resample]: https://output.circle-artifacts.com/output/job/fe5bb357-984b-4745-99a8-decbcc21ed8a/artifacts/0/html/reference/generated/scipy.signal.resample.html
[decimate]: https://output.circle-artifacts.com/output/job/fe5bb357-984b-4745-99a8-decbcc21ed8a/artifacts/0/html/reference/generated/scipy.signal.decimate.html

DietBru committed:
MAINT: signal.resample/signal.envelope: Port functions to array API standard.

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3500,38 +3506,58 @@ def invresz(r, p, k, tol=1e-3, rtype='avg'):
     t : array_like, optional
-        If `t` is given, it is assumed to be the equally spaced sample
-        positions associated with the signal data in `x`.
+        If `t` is not ``None`` (default), then the timestamps of the resampled signal
```
Minor, trivial: this got me pause for a moment, to parse whether the default is None or not None. Consider adding the "Default is `None`" at the end instead.

DietBru replied:
Fixed in commit 76636f9e7

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3500,38 +3506,58 @@ def invresz(r, p, k, tol=1e-3, rtype='avg'):
-        ``freq`` Consider the input `x` as frequency-domain.
+        If not ``None`` (default), specifies a filter in the Fourier domain, which is
+        applied before resampling. I.e., the FFT ``X`` of `x` is calculated by
+        ``X = W * fft(x, axis=axis)``. ``W`` may be interpreted as a sepctral
```
```suggestion
        ``X = W * fft(x, axis=axis)``. ``W`` may be interpreted as a spectral
```

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3500,38 +3506,58 @@ def invresz(r, p, k, tol=1e-3, rtype='avg'):
+        ``f_X = fftfreq(n_x, T)``.
+
+        If `window` is a 1d array of length `n_x` then ``W=window``.
+        If `window` is a function  then ``W = window(f_X)``.
```
```suggestion
        If `window` is a callable  then ``W = window(f_X)``.
```

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3500,38 +3506,58 @@ def invresz(r, p, k, tol=1e-3, rtype='avg'):
+        If `window` is a 1d array of length `n_x` then ``W=window``.
+        If `window` is a function  then ``W = window(f_X)``.
+        Otherwise, `window` is passed to `~scipy.signal.get_window`, i.e.,
+        ``W = fftshift(signal.get_window(window, n_x))``.
```
Just to double-check: is `fftshift` not extraneous here?

DietBru replied:
It is not, since it is used in the code as well. The reason is  that you typically want your spectral window to be centered over the zero frequency bin. For the standard `fft()`, the first entry corresponds to the zero frequency bin, hence `fftshift` is needed.

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3540,170 +3566,152 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
+    illustrated in the example below) can result in significant speed increases.
+
+    `resample` is intended to be used for periodic signals with equidistant sampling
+    intervals. For non-periodic signals, `resample_poly` may be a better choice.
```
Consider stressing this in the See Also link to `resample_poly`.

DietBru replied:
Added in commit 76636f9e7.

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3540,170 +3566,152 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
 
-    Due to ``N`` being prime,
+    The following example illustrates that `~scipy.fft.rfft` / `~scipy.fft.irfft`
+    combination does not always produce the correct resampling result:
```
Consider 

```suggestion
    combination does not always produce the correct resampling result, while `resample` does:
```

or something to this effect.

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3540,170 +3566,152 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
+    x1 = [ 2. -2.  2. -2.]
+    y1 = [ 1. -1.  1. -1.]
+
+    It can be easily verified that ``x1`` is correct and ``y1`` is not. To produce
```
Consider showing how to verify it.

If the docstring gets too long, consider moving some material into the tutorial, and linking from here.

DietBru replied:
Reworked example in commit 76636f9e7.

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3540,170 +3566,152 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
+    >>> x1 = resample(x0, num=n1)
+    >>> y1 = irfft(rfft(x0), n=n1) * n1/n0
+    >>> for x_, n_ in zip((x0, x1, y1), ('x0', 'x1', 'y1')):
+    ...    print(f"{n_} = {x_}")
```
Optionally, consider simplifying to

```
>>> x0
<copy-paste the output>
>>> x1
<copy-paste the output>
>>> y1
<copy-paste the output>
```

If you prefer the current version for readability, fine.
The suggestion makes it a little friendlier to the docstring checker (it has some logic to parse printed arrays from text, but here it won't manage and will fall back to literal symbol-by-symbol comparison.

DietBru replied:
Good point. Fixed in commit 76636f9e7.

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3540,170 +3566,152 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
 
-    >>> f = signal.resample(y, num)
+    Note that the `y1` and `y1_p` are not identical due to the differences in the
+    utilized antialiasing filter.
```
This gets me confused a bit. (Maybe it's very very obvious for DSP regulars, but).

Results are not identical, OK. As a user, do I need to worry about it:
Is the difference just fp noise? Is the difference significant? Is there a way to make them identical? Is there actually a need to make them identical?

DietBru replied:
Well, you should always use an anti-aliasing filter, when resampling. The default for `resample`, i.e. `window=None`, corresponds to passing `np.ones(len(x))`. To replicate the default behavior of `resample_poly` you would have to pass something along the lines of `window=fft(firwin(.., window=('kaiser', 5.0)), n=len(x))`.
To reproduce identical behavior with `resample_poly`, the filter length would have to be same as the signal length, which would defeat the point of using it.

I noticed while rereading that in the snippet above that `resample_poly` is missing the parameter `padtype='wrap' to account for the periodicity of the signal. I'll fix that in the next commit.

ev-br replied:
Consider adding this kind of discussion to the tutorial and linking from this docstring.

DietBru replied:
In commit 76636f9e7, I added a pointer to the relationship between the `window` parameter and the anti-aliasing filter.

I have never thought about how to explain within two or three paragraphs what is going on spectrally when resampling. This topic deems me to demand at least two to three pages in a well written text book. Unfortunately, I am not aware of any good English references (learned this stuff too long ago...). In summary, I am unsure how I would approach this in a tutorial. I am very open to suggestions for references or to approaches for concise explanations.

ev-br replied:
I've to admit  https://github.com/scipy/scipy/pull/22814#discussion_r2039204850 is clearer than either of the docstring additions, to me at least. Consider just this explanation somewhere, here or in a follow-up.
I don't have more specific suggestions though, so up to you.

DietBru replied:
Hmm, it's interesting how subjective perspective differ. What my explanation are lacking, is i) what the purpose of the anti-aliasing filter serves and ii) what criteria it needs to fulfill. They key (of course, from my biased point of view) lies in understanding how sampled spectra are periodic with an interval width of 1 / sampling interval...  Would a link to this [Wikipedia section] on sampling & DFT be in some way enlightening (imhoi, it's slightly besides the key points)? A more complete treatment can be found in Sec. 2.4 and 2.5 of  [Vetterli's book] (manuscript is online—imho, it's a bit too compact for newcomers)

Would it be a possibility, if you would formulate in a new issue what you think is worthwhile explaining in a tutorial section. The process of resampling, aliasing, ...?

[Wikipedia section]: https://en.wikipedia.org/wiki/Discrete-time_Fourier_transform#Sampling_the_DTFT
[Vetterli's book]: http://waveletsandsubbandcoding.org/

ev-br commented on scipy/signal/_signaltools.py:
```diff
@@ -3540,170 +3566,152 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
-        return y, new_t
+        W = sp_fft.fftshift(get_window(window, n_x))
+
+    if domain == 'time' and xp.isdtype(x.dtype, 'real floating'):  # utilize rfft():
```
So integer signals go through the two-sided FFT code path, even in time domain, is this intentional?

DietBru replied:
Good catch! Addressed in commit 76636f9e7.

ev-br reviewed (commented):
Looks very reasonable overall from a non-specialist POV.
Cannot comment on fine DSP details, am ready to trust your judgement.

I've left several comments, all of them minor and mostly optional.

DietBru committed:
Apply suggestions from code review

Co-authored-by: Evgeni Burovski <evgeny.burovskiy@gmail.com>

DietBru commented on scipy/signal/_signaltools.py:
```diff
@@ -2893,12 +2894,12 @@ def envelope(z: np.ndarray, bp_in: tuple[int | None, int | None] = (1, None), *,
         Z = sp_fft.fft(z)
     else:  # avoid calculating negative frequency bins for real signals:
-        Z = np.zeros_like(z, dtype=sp_fft.rfft(z.flat[:1]).dtype)
+        Z = xp.zeros_like(z, dtype=sp_fft.rfft(z.flat[:1]).dtype)
```
IMO, the `fft` function does have some surprising behavior in that regard:
```python
In [1]: import numpy as np

In [2]: from scipy.fft import fft

In [3]: fft(np.array([1,], dtype=np.float32)).dtype
Out[3]: dtype('complex64')

In [4]: fft(np.array([1,], dtype=np.int16)).dtype
Out[4]: dtype('complex128')
```
Furthermore, you can [swap in] your own back-end, if you want.

[swap in]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.set_backend.html

ev-br replied:
Ok got it. Best kept as is then, agreed.

DietBru committed:
MAINT: signal.resample: Improve docstr and apply review suggestions

DietBru commented:
Thanks a lot for the detailed review, @ev-br! I hope I addressed all your points.

Rendered docs: [resample], [decimate]

[resample]: https://output.circle-artifacts.com/output/job/2db5a33b-bd5e-4b92-8f31-fd4de048a07e/artifacts/0/html/reference/generated/scipy.signal.resample.html
[decimate]: https://output.circle-artifacts.com/output/job/2db5a33b-bd5e-4b92-8f31-fd4de048a07e/artifacts/0/html/reference/generated/scipy.signal.decimate.html

ev-br reviewed (approved):
Looks good from a non-specialist perspective.

Let's keep this open for a short while, in a hope that one of DSP experts looks through the PR. 
Otherwise, I think it's good to go in.

DietBru commented:
:partying_face: Merge whenever you think it's appropriate. Thanks again for your helpful input, @ev-br!

lucascolley commented:
> Looks good from a non-specialist perspective.
> 
> Let's keep this open for a short while, in a hope that one of DSP experts looks through the PR.

If you have some time @endolith @larsoner !

larsoner commented on scipy/signal/_signaltools.py:
```diff
@@ -3500,210 +3506,230 @@ def invresz(r, p, k, tol=1e-3, rtype='avg'):
+
+    Note that the `y1` and `y1_p` are not identical due to the differences in the
+    utilized antialiasing filter in each function. In both functions the antialiasing
+    filter can be customized by modifying the `window` parameter, though `resample`
+    interprets it as a spectral window function, whereas `resample_poly` assumes it to
+    be some form of impulse response.
```
I would say something like this... the "some form of impulse response" is what caught my eye as being unclear. The purpose in `resample_poly` is to perform the antialiasing. If you pass a bad antialiasing filter, you risk aliasing there. There is no risk of aliasing in `resample` since the ideal brick-wall filter is always implicitly applied by the method itself.
```suggestion
    filter can be customized by modifying the `window` parameter. `resample`
    interprets it as a *spectral* window function that is always applied in addition to
    the implicit ideal (brick wall) antialiasing filter. `resample_poly` applies
    `window` as a FIR filter that gets applied in the *time* domain and should be
    designed appropriately to avoid aliasing.
```

DietBru replied:
Addressed in commit [b3c1993](https://github.com/scipy/scipy/pull/22814/commits/b3c199349fd5cc424dbbaa77f3b615881156a666)

larsoner reviewed (commented):
I like the docstring updates! The code changes are extensive but I had a look at them and they looked okay, and I think we can trust the existing tests here. Just a couple of comments/ideas for the docs then I think we're good to go

ev-br commented:
gentle ping @DietBru -- would you be open to addressing Eric's comments in this PR or a follow-up, or you two are going to speak some DSP and decide to not implement his suggestions?

DietBru committed:
MAINT: signal.resample: Improve docstr examples

[docs only]

DietBru commented:
Sorry for not being able to finish this sooner. In commit [b3c1993](https://github.com/scipy/scipy/pull/22814/commits/b3c199349fd5cc424dbbaa77f3b615881156a666), I added two additional plots to the resample examples. I hope this addresses all your review remarks, @larsoner and perhaps address some aspects of this [comment], @ev-br.


Rendered docs: [resample], [decimate]

[comment]: https://github.com/scipy/scipy/pull/22814#discussion_r2037119941
[resample]: https://output.circle-artifacts.com/output/job/609d1aa8-ab9c-4da7-9076-ef7843219850/artifacts/0/html/reference/generated/scipy.signal.resample.html
[decimate]: https://output.circle-artifacts.com/output/job/609d1aa8-ab9c-4da7-9076-ef7843219850/artifacts/0/html/reference/generated/scipy.signal.decimate.html

ev-br commented:
Let's land this. Thanks @DietBru , @larsoner !
